controllers/baseboardmanagement_controller.go

```diff
@@ -18,42 +18,193 @@ package controllers
 }
 
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=baseboardmanagements,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=baseboardmanagements/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=baseboardmanagements/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=secrets;,verbs=get;list;watch
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the BaseboardManagement object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile ensures the state of a BaseboardManagement.
+// Gets the BaseboardManagement object and uses the SecretReference to initialize a BMC Client.
+// Ensures the BMC power is set to the desired state.
+// Updates the status and conditions accordingly.
 func (r *BaseboardManagementReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx).WithValues("BaseboardManagement", req.NamespacedName)
+	logger.Info("Reconciling BaseboardManagement", "name", req.NamespacedName)
 
-	// TODO(user): your logic here
+	// Fetch the BaseboardManagement object
+	baseboardManagement := &bmcv1alpha1.BaseboardManagement{}
+	err := r.Client.Get(ctx, req.NamespacedName, baseboardManagement)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BaseboardManagement", "name", req.NamespacedName)
+		return ctrl.Result{}, fmt.Errorf("failed to get BaseboardManagement: %v", err)
+	}
+
+	// Deletion is a noop.
+	if !baseboardManagement.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	return r.reconcile(ctx, baseboardManagement)
+}
+
+func (r *BaseboardManagementReconciler) reconcile(ctx context.Context, bm *bmcv1alpha1.BaseboardManagement) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("BaseboardManagement", bm.Name)
+
+	// Fetching username, password from SecretReference
+	username, password, err := r.resolveAuthSecretRef(ctx, bm.Spec.Connection.AuthSecretRef)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("resolving authentication from SecretReference: %v", err)
+	}
+
+	// Initializing BMC Client connection
+	r.BMCClient.InitClient(bm.Spec.Connection.Host, "623", username, password)
+	err = r.BMCClient.OpenConnection(ctx)
+	if err != nil {
+		logger.Error(err, "BMC connection failed", "host", bm.Spec.Connection.Host)
+		result, setConditionErr := r.setCondition(ctx, bm, bmcv1alpha1.ConnectionError, err.Error())
```

Comment:
Sure, I was thinking on this. We could add condition types, like `PowerStatusFailed` `PowerSetFailed`?

Reply:
Are we sure we have conditions right? A condition is typically defining a particular condition about the current state, for example `Connected`, not a piece of data. That doesn't mean we can't add supplementary data like the error message but the condition itself needs to make sense and I'm not sure this `ConnectionError` does.

I feel like the following makes more sense

```go
const (
   Connected = "Connected"
)

type BaseboardManagementCondition struct {
	// Type of the BaseboardManagement condition.
	Type BaseboardManagementConditionType `json:"type"`

    // Could be True, False or Unknown (empty string).
	Status BaseBoardManagementConditionStatus `json:"status"`

	// Message represents human readable message indicating details about last transition.
	// +optional
	Message string `json:"message,omitempty"`
}
```

In the case of `Connected` a status of `False` indicates a problem and an error message is likely available where as `Unknown` indicates we haven't even attempted to connect.

Reply:
With respect to communicating a power status action failed that's really an event. We've tried to reconcile status by retrieving the current power state (or setting it) and it was unsuccessful so we generate an event for an operator to inspect. Events show up in the `kubectl describe` output.

I mention that because we shouldn't get carried away with conditions. Conditions are meant to be useful for _machines_, events should be useful for _humans_.